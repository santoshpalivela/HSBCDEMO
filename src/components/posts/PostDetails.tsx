import React, { useEffect, useState } from "react";
import { RouteComponentProps } from "react-router-dom";
import axios from "axios";
import { Post } from "../../types";
import "../../scss/postdetails.scss";
import Loader from "react-loader-spinner";
import { toast } from "react-toastify";

interface Props extends RouteComponentProps {}
export const PostDetails: React.FC<Props> = (props: Props) => {
  const [details, setDetails] = useState<Post>({
    userId: 1,
    id: 0,
    title: "",
    body: "",
  });
  const [showloader, setShowLoader] = useState<boolean>(false);

  const handleGoBack = (): void => {
    props.history.push(`/posts`);
  };

  useEffect(() => {
    setShowLoader(true);
    if (
      props.match &&
      props.match.params &&
      props.match.params.hasOwnProperty("id")
    ) {
      let id = props.match.params["id"];
      axios
        .get(`https://jsonplaceholder.typicode.com/posts/${id}`)
        .then(function (response) {
          setDetails(response.data);
          setShowLoader(false);
        })
        .catch(function (error) {
          toast.error(error.message);
          setShowLoader(false);
        });
    }
  }, []);
  return (
    <div className="container">
      <div>
        <button className="btn-goback" onClick={handleGoBack}>
          {" "}
          Go back{" "}
        </button>
      </div>
      <h2>Details</h2>
      {showloader ? (
        <div className="div-spinner">
          <Loader type="Oval" color="#000" height={100} width={100} />
        </div>
      ) : (
        <div>
          <div className="lbl-details">
            <span className="lbl-bold"> UserId : </span>
            <span>{details.userId}</span>
          </div>
          <div className="lbl-details">
            <span className="lbl-bold"> Id : </span>
            <span>{details.id}</span>
          </div>
          <div className="lbl-details">
            <span className="lbl-bold"> Title : </span>
            <span>{details.title}</span>
          </div>
          <div className="lbl-details">
            <span className="lbl-bold"> Body : </span>
            <span>{details.body}</span>
          </div>
        </div>
      )}
    </div>
  );
};
