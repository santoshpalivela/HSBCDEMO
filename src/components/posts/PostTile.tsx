import React from "react";
import {Post} from "../../types";
// import "../../scss/posttile.scss";

import "../../scss/posttile.scss"
export const PostTile:React.FC<Post>=({userId,title}:Post)=>{
   
    return (
    <div className="tile-card">
        
        <div className="lbl-userid">
            <span >UserId:</span>
            <span className="userid-val" >{userId}</span>

        </div>
        <div className="lbl-title">
            <span>Title:</span>
            <span className="title-val">{title}</span>
        </div>
    </div>
    );
    }