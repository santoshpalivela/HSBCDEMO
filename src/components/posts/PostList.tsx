import React, { useEffect, useState } from "react";
import { RouteComponentProps } from "react-router-dom";
import { Post } from "../../types";
import { PostTile } from "./PostTile";
import axios from "axios";
import { toast } from "react-toastify";
import "../../scss/postlist.scss";

interface Props extends RouteComponentProps {}

export const PostList: React.FC<Props> = (props: Props) => {
  const [postList, setPostList] = useState<Post[]>([]);
  const [filteredList, setFilteredList] = useState<Post[]>([]);
  const [searhTerm, setSearchTerm] = useState<string>("");

  const onTileClick = (id: number) => {
    props.history.push(`/details/${id}`);
  };

  const onSearch = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const searhTerm = e.target.value;
    let filteredPosts;

    if (searhTerm === "") filteredPosts = postList;
    else
      filteredPosts = postList.filter(
        (item) =>
          item.title.indexOf(searhTerm) != -1 ||
          item.body.indexOf(searhTerm) != -1
      );

    setSearchTerm(searhTerm);
    setFilteredList(filteredPosts);
  };

  const handleLogout = () => {
    sessionStorage.clear();
    props.history.push("/");
  };

  const fetchDelayedPosts = (sortedData: Post[], iteration: number): Post[] => {
//     if (searhTerm !== "") 
//     {
// debugger;
//       return [];
//     }
//     else
       return sortedData.slice(0, iteration + 1);
  };

  useEffect(() => {
    axios
      .get("https://jsonplaceholder.typicode.com/posts")
      .then(function (response) {
           const sortedData = response.data.sort(function (a, b) {
          return b.id - a.id;
        });

        const itemsCount = sortedData.length;

        for (let i = 0; i < itemsCount-10; i++) {
          setTimeout(() => {
            const newData = fetchDelayedPosts(sortedData, i);
            
            if (newData.length > 0) {
              setPostList(newData);
              setFilteredList(newData);
            }
          }, i * 1000);
        }
      })
      .catch(function (error) {
           toast.error(error.message);
        });
  }, []);

  return (
    <div className="container">
      <div className="row search-bar">
        <button className="btn-logout" onClick={handleLogout}>
          Logout
        </button>
        <input
          className="txt-search"
          onChange={onSearch}
          placeholder="search"
        />
      </div>
      <div className="post-list-container">
        {filteredList && filteredList.length > 0 ? (
          filteredList.map((item ,index) => (
            <div
              key={item.id}
              onClick={() => {
                onTileClick(item.id);
              }}
              className="col-lg-4"
            >
              <PostTile {...item} />
            </div>
          ))
        ) : (
          <div className="no-data">No Records found </div>
        )}
      </div>
    </div>
  );
};
