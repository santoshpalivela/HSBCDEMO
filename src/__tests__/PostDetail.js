import React from "react";

import { shallow, mount, render, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import {PostDetails} from "../components/posts/PostDetails";
import renderer from "react-test-renderer";



configure({ adapter: new Adapter() });

describe("should render PostDetailsMounted without throwing an error", () => {

  
  const tree = renderer.create(<PostDetails />).toJSON();
  expect(tree).toMatchSnapshot();

  const mockFn = jest.fn();

  const PostDetailsMounted = shallow(
    <PostDetails />
  );

  it("Find the  root of post details component ", () => {
    expect(PostDetailsMounted.find(".container").exists()).toBe(true)

  });

  it("Find the  Goback button  ", () => {
    expect(PostDetailsMounted.find(".btn-goback").exists()).toBe(true)

    // PostDetailsMounted.find(".btn-goback").simulate("click")
  });





});
