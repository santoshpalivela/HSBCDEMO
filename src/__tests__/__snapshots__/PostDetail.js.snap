// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="container"
>
  <div>
    <button
      className="btn-goback"
      onClick={[Function]}
    >
       
      Go back
       
    </button>
  </div>
  <h2>
    Details
  </h2>
  <div>
    <div
      className="lbl-details"
    >
      <span
        className="lbl-bold"
      >
         UserId : 
      </span>
      <span>
        1
      </span>
    </div>
    <div
      className="lbl-details"
    >
      <span
        className="lbl-bold"
      >
         Id : 
      </span>
      <span>
        0
      </span>
    </div>
    <div
      className="lbl-details"
    >
      <span
        className="lbl-bold"
      >
         Title : 
      </span>
      <span>
        
      </span>
    </div>
    <div
      className="lbl-details"
    >
      <span
        className="lbl-bold"
      >
         Body : 
      </span>
      <span>
        
      </span>
    </div>
  </div>
</div>
`;
